### PR TITLE
MAILBOX-314 search: precise spec about search on email and fix ES impl

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[7d52e7fbb94b28b292047c54ef5c04c52ab9d59e] Failed tests: 
[7d52e7fbb94b28b292047c54ef5c04c52ab9d59e]   GetMessageListMethodTest.getMessageListShouldSortMessagesWhenSortedBySubjectAsc:1153 1 expectation failed.
[7d52e7fbb94b28b292047c54ef5c04c52ab9d59e] JSON path [0][1].messageIds doesn't match.
[7d52e7fbb94b28b292047c54ef5c04c52ab9d59e] Expected: iterable containing ["3bc76fe0-30b5-11e8-a05b-fd1fb17f40eb", "3bcaa430-30b5-11e8-a05b-fd1fb17f40eb"]
[7d52e7fbb94b28b292047c54ef5c04c52ab9d59e]   Actual: [3bcaa430-30b5-11e8-a05b-fd1fb17f40eb, 3bc76fe0-30b5-11e8-a05b-fd1fb17f40eb]
[7d52e7fbb94b28b292047c54ef5c04c52ab9d59e] 
[7d52e7fbb94b28b292047c54ef5c04c52ab9d59e]   GetMessageListMethodTest.getMessageListShouldSortMessagesWhenSortedBySubjectDesc:1175 1 expectation failed.
[7d52e7fbb94b28b292047c54ef5c04c52ab9d59e] JSON path [0][1].messageIds doesn't match.
[7d52e7fbb94b28b292047c54ef5c04c52ab9d59e] Expected: iterable containing ["30cee780-30b5-11e8-a05b-fd1fb17f40eb", "30cb3e00-30b5-11e8-a05b-fd1fb17f40eb"]
[7d52e7fbb94b28b292047c54ef5c04c52ab9d59e]   Actual: [30cb3e00-30b5-11e8-a05b-fd1fb17f40eb, 30cee780-30b5-11e8-a05b-fd1fb17f40eb]
```

[Investigating]

**Comment by chibenwa** [comment]
test this please

(never knowns)

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 1]
```diff
@@ -766,6 +795,26 @@ public void addressShouldReturnUidHavingRightExpeditorWhenFromIsSpecified() thro
             .containsOnly(m8.getUid());
     }
 
+    @Test
+    public void addressShouldReturnUidHavingRightExpeditorWhenFromIsSpecifiedWithOnlyUserPartOfEmail() throws Exception {
+        SearchQuery searchQuery = new SearchQuery(SearchQuery.address(AddressType.From, "murari"));
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: you don't need:
```
+        Assume.assumeTrue(storeMailboxManager
+            .getSupportedSearchCapabilities()
+            .contains(MailboxManager.SearchCapabilities.PartialEmailMatch));
```
here?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 2]
```diff
@@ -774,9 +823,66 @@ public void addressShouldReturnUidHavingRightRecipientWhenToIsSpecified() throws
             .containsOnly(m1.getUid());
     }
 
+    @Test
+    public void addressShouldReturnUidHavingRightRecipientWhenToIsSpecifiedWithOnlyEmailUserPart() throws Exception {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: idem?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 3]
```diff
@@ -774,9 +823,66 @@ public void addressShouldReturnUidHavingRightRecipientWhenToIsSpecified() throws
...
     @Test
     public void addressShouldReturnUidHavingRightRecipientWhenCcIsSpecified() throws Exception {
-        SearchQuery searchQuery = new SearchQuery(SearchQuery.address(AddressType.Cc, "any@any.com"));
+        SearchQuery searchQuery = new SearchQuery(SearchQuery.address(AddressType.Cc, "monkey@any.com"));
+        assertThat(messageSearchIndex.search(session, mailbox, searchQuery))
+            .containsOnly(m5.getUid());
+    }
+
+    @Test
+    public void addressShouldReturnUidHavingRightRecipientWhenCcIsSpecifiedWithOnlyUserPartOfTheEmail() throws Exception {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: idem?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java [hunk 4]
```diff
@@ -774,9 +823,66 @@ public void addressShouldReturnUidHavingRightRecipientWhenToIsSpecified() throws
...
+            .getSupportedSearchCapabilities()
+            .contains(MailboxManager.SearchCapabilities.PartialEmailMatch));
+
+        SearchQuery searchQuery = new SearchQuery(SearchQuery.address(AddressType.Cc, "any.com"));
+        assertThat(messageSearchIndex.search(session, mailbox, searchQuery))
+            .containsOnly(m5.getUid());
+    }
+
+    @Test
+    public void addressShouldReturnUidHavingRightRecipientWhenBccIsSpecifiedWithOnlyUserPartOfTheEmail() throws Exception {
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: idem?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/search/AbstractMessageSearchIndexTest.java]: I will add ...

**Comment by mbaechler** [comment]
merged

**Pull request closed without merging**